File: stack/simplestack/simplestack.py

'''

	Note : Array in python can be used as stack

	Time copmlexity :

		access    : O(n)
		delete    : O(1)
		search    : O(n)
		insertion : O(1) 

'''

class stack :

	def __init__(self) :

		self.array = []

	def _empty(self) :

		return True if len(self.array) == 0 else False

	def _push(self , value) :

		self.array.append(value)

	def _pop(self) :

		if self._empty() :
			print("Empty Stack")
		else :
			return self.array.pop()

	def _peek(self) :

		return self.array[-1] if not self._empty() else "Error : Empty Stack"

	def _size(self) :

		return len(self.array)

	def _display(self) :

		if self._empty() :
			print("Empty Stack")
		else :
			while not self._empty() :
				print(self._peek())
				self._pop()

s = stack()

s._push(1) ;
s._push(2) ;
s._pop() ;
s._push(3) ;
s._push(4) ;
s._pop() ;
s._push(5) ;
s._push(6) ;
s._push(7) ;
s._pop() ;

print(f"Size : {s._size()}")

s._display() ;

File: stack/simplestack/simplestack.js

/*

	Note : Array in Javascript can be used as stack

	Time copmlexity :

		access    : O(n)
		delete    : O(1)
		search    : O(n)
		insertion : O(1) 

*/

class stack
{
	constructor()
	{
		this.array = [] ;
		this.top = -1 ;
	}

	_empty()
	{
		if(this.top === -1)
			return true ;
		return false ;
	}

	_push(value)
	{
		this.top++ ;
		this.array[this.top] = value ;

	}

	_pop()
	{
		if(this._empty())
		{
			console.log("Empty Stack") ;
		}
		else
		{
			this.top-- ;
		}
	}

	_peek()
	{
		return this.array[this.top] ;
	}

	_display()
	{
		var c = this.top ;
		while(c >= 0)
		{
			console.log(this.array[c]) ;
			c-- ;
		}
	}

	_length()
	{
		return this.top + 1 ;
	}
}

var s = new stack() ;

s._push(1) ;
s._push(2) ;
s._pop() ;
s._push(3) ;
s._push(4) ;
s._pop() ;
s._push(5) ;
s._push(6) ;
s._push(7) ;
s._pop() ;
s._display() ;

console.log("Size : " + s._length()) ;
